src/rdma.c

```diff
@@ -0,0 +1,1894 @@
+        goto reject;
+    }
+
+    ret = rdma_accept(cm_id, &conn_param);
+    if (ret) {
+        serverRdmaError(err, "RDMA: accept failed");
+        goto free_rdma;
+    }
+
+    return C_OK;
+
+free_rdma:
+    rdmaReleaseResource(ctx);
+reject:
+    /* reject connect request if hitting error */
+    rdma_reject(cm_id, NULL, 0);
+
+    return C_ERR;
+}
+
+static rdma_listener *rdmaFdToListener(connListener *listener, int fd) {
+    for (int i = 0; i < listener->count; i++) {
+        if (listener->fd[i] != fd) continue;
+
+        return (rdma_listener *)listener->priv + i;
+    }
+
+    return NULL;
+}
+
+/*
+ * rdmaAccept, actually it works as cm-event handler for listen cm_id.
+ * accept a connection logic works in two steps:
+ * 1, handle RDMA_CM_EVENT_CONNECT_REQUEST and return CM fd on success
+ * 2, handle RDMA_CM_EVENT_ESTABLISHED and return C_OK on success
+ */
+static int rdmaAccept(connListener *listener, char *err, int fd, char *ip, size_t ip_len, int *port, void **priv) {
+    struct rdma_cm_event *ev;
+    enum rdma_cm_event_type ev_type;
+    int ret = C_OK;
+    rdma_listener *rdma_listener;
+
+    rdma_listener = rdmaFdToListener(listener, fd);
+    if (!rdma_listener) {
+        serverPanic("RDMA: unexpected listen file descriptor");
+    }
+
+    ret = rdma_get_cm_event(rdma_listener->cm_channel, &ev);
+    if (ret) {
+        if (errno != EAGAIN) {
+            serverLog(LL_WARNING, "RDMA: listen channel rdma_get_cm_event failed, %s", strerror(errno));
+            return ANET_ERR;
+        }
+        return ANET_OK;
+    }
+
+    ev_type = ev->event;
+    switch (ev_type) {
+    case RDMA_CM_EVENT_CONNECT_REQUEST:
+        ret = rdmaHandleConnect(err, ev, ip, ip_len, port);
+        if (ret == C_OK) {
+            RdmaContext *ctx = (RdmaContext *)ev->id->context;
+            *priv = ev->id;
+            ret = ctx->comp_channel->fd;
+        }
+        break;
+
+    case RDMA_CM_EVENT_ESTABLISHED: ret = rdmaHandleEstablished(ev); break;
+
+    case RDMA_CM_EVENT_UNREACHABLE:
+    case RDMA_CM_EVENT_ADDR_ERROR:
+    case RDMA_CM_EVENT_ROUTE_ERROR:
+    case RDMA_CM_EVENT_CONNECT_ERROR:
+    case RDMA_CM_EVENT_REJECTED:
+    case RDMA_CM_EVENT_ADDR_CHANGE:
+    case RDMA_CM_EVENT_DISCONNECTED:
+    case RDMA_CM_EVENT_TIMEWAIT_EXIT:
+        rdmaHandleDisconnect(ev);
+        ret = C_OK;
+        break;
+
+    case RDMA_CM_EVENT_MULTICAST_JOIN:
+    case RDMA_CM_EVENT_MULTICAST_ERROR:
+    case RDMA_CM_EVENT_DEVICE_REMOVAL:
+    case RDMA_CM_EVENT_ADDR_RESOLVED:
+    case RDMA_CM_EVENT_ROUTE_RESOLVED:
+    case RDMA_CM_EVENT_CONNECT_RESPONSE:
+    default: serverLog(LL_NOTICE, "RDMA: listen channel ignore event: %s", rdma_event_str(ev_type)); break;
+    }
+
+    if (rdma_ack_cm_event(ev)) {
+        serverLog(LL_WARNING, "ack cm event failed\n");
+        return ANET_ERR;
+    }
+
+    return ret;
+}
+
+static void connRdmaAcceptHandler(aeEventLoop *el, int fd, void *privdata, int mask) {
```

Comment:
How many connections can this support?

Valkey is often used with very large numbers of connections.  The documentation promises that Valkey is able to handle 10,000.  According to file:///C:/Users/z/Downloads/icnp21camera-paper30.pdf from 2021, "However, the performance of RDMA degrades as the number of connections grows. Our testbed measurements (Fig. 1) show that even the latest RNIC can only support less than 450 concurrent connections if to maintain the peak performance (detailed settings described in §II-B)."  Here is a graph from that paper.

![image](https://github.com/valkey-io/valkey/assets/76451671/38883851-98e1-4243-923f-b42219203686)

If the limit is still that severe today, could we get around it by supporting multiple RNICs?

Reply:
This command can be used to query the max number of QP supported by hardware:
`ibv_devinfo -v | grep max_qp -w`

For example, mellanox cx5 supports 262144. But yes, the paper shows the performance drop when lots of QP run heavy throughput together.
However, in the real scenario, valkey-server is not a throughput oriented server, large key-value is not friendly which increases the latency significantly. Too many active connections also leads significant latency. As far as I see, Valkey over RDMA provides a higher performance platform with less connections(also more hardware aware knowledge).

And the current version supports multiple RNICs, for example, `--loadmodule src/valkey-rdma.so port=6379 bind=*` will listen on all the RNICs.

Reply:
> This command can be used to query the max number of QP supported by hardware: `ibv_devinfo -v | grep max_qp -w`
> 
> For example, mellanox cx5 supports 262144.

The question is not "How many QPs does the RNIC support?"  The question is "How many connection states can be retained on the RNIC at the same time?"  The graph shown above was the results of testing with a "Mellanox ConnectX-6 Dx EN NIC".

The RNIC is caching connection states.  If the size of the working set grows larger than the cache, thrashing sets in.  Support for 10,000 clients is a common claim for Valkey (sorry, I can't find a link that corresponds to [Redis client handling](https://redis.io/docs/latest/develop/reference/clients/)).  I have seen [Amazon](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/BestPractices.Clients.Redis.Connections.html) claim to support "up to 65,000 concurrent client connections."

This PR remains viable so long as the number of active connections via a single RNIC remains small (< ~ 450 for a Mellanox ConnectX-6 Dx EN NIC).

It seems to me that aggressive testing with large numbers of clients is needed.  Otherwise, it is up to the user to experimentally determine if their configuration is viable.

Reply:
Hi, I'm not going to test this pattern, and I don't care this scenario. If a server is going to accept lots of connections, using TCP instead would be a good choice. Depending on the workload, separating valkey instance into 'RDMA friendly' and 'RDMA unfriendly' types, it's quite easy to select TCP or RDMA.
Just like, a user want to enable disk saving, he needs to know HDD, SSD or NVMe is in use.

Reply:
I think you can also use SRQ to handle the scale better

Reply:
> I think you can also use SRQ to handle the scale better

Right, the feature is in the plan. I'd like to add a new config option like 'srq-wr', if 0, use regular RQ for QP, otherwise use SRQ instead. Because of the initial version, I don't want to make it complex.